### DISCUSSION TITLE
Is it possible to build the WASM native libraries on a Windows pc?

### DISCUSSION THREAD

**Discussion by BradChase2011** [mono/SkiaSharp#3024] [Q&A] [answered]
We have changes to skia that we need built into the native libs.  We can build all the end points we have except for WASM.  It skips building the native libs and if I am looking at the correct cake, it seems to drop off on the task libSkiaSharp with ".WithCriteria(IsRunningOnLinux())".

Is it possible to build the wasm libs on a windows pc?  If so, what command should I use to test it out?  If not, is there a document or any info on how to build them?

Thanks much!
Brad

**Comment by mattleibow** [comment, marked as answer]
You should be able to use WSL as that is what I use. Or Docker.

In the scripts/Docker folder, the things are there and there is a build local script you can run to do it all for you.

I am not sure if you can build wasm on windows, I just never tried and then never needed to. Maybe it will work if you remove the condition?

**Reply by BradChase2011** [reply]
I tried removing it but it hit error after error so I moved on from that thinking it most likely was purposefully put there.  I tried the WSL route, which I would prefer but I am not sure what to install and what versions to install.  For instance the WASM side is built for different versions of emscripten but I cannot find out which versions it needs built for or where a script is that tells it so.  

I didnt even recognize the docker route, so I went ahead and ran that and it did build.  That said, What do I do with the files?  The script says 3.1.34 which is the version of emscripten we are on currently, so thats ok for now.  I guess the question is, where do I place them to build the nuget packges?  With the download externals it places them in a directory structure under the externals, ".\SkiaSharp\output\native\wasm\", and then it looks like so:

![image](https://github.com/user-attachments/assets/3a6d0bd6-4703-4323-bfd1-37af3c7c348e)

![image](https://github.com/user-attachments/assets/a96b49b6-292a-4cfb-984d-07a140a30c8c)

![image](https://github.com/user-attachments/assets/8ae30fcf-2fbd-41e6-a88b-2bf68e4ba7cc)

I take it for the libskiasharp, we copy it from .\SkiaSharp\externals\skia\out\wasm\libSkiaSharp.a over to .\SkiaSharp\output\native\wasm\libSkiaSharp.a\3.1.34\" do we just replace all them with that one library?  Copy over the SIMDs, multi threaded and single threaded assemblies?


EDIT:  I went ahead and added the libSkiaSharp.a and libHarfBuzzSharp.a into the SkiaSharp\output\native\wasm directory and it appears in the native assets packages under the buildTransitive\netstandard1.0 directories but at runtime I still get "[ERROR] FATAL UNHANDLED EXCEPTION: System.DllNotFoundException: libSkiaSharp".  I can try some more rebuilds and package clearing but I am concerned that it wont be the correct library under the different variations of emscripten.

EDIT2:  I went ahead and added the libSkiaSharp.a and libHarfBuzzSharp.a into the SkiaSharp\output\native\wasm\libHardBuzzSharpo.a\3.1.34\st\ directory and SkiaSharp\output\native\wasm\libSkiaSharp.a\3.1.34\st\ directory after looking at the target and props files in the nuget.  This seemingly confirms my fears that I need to build all of them.  I know soon we will be moving forward from emscripten 3.1.34.  Unfortunately I still get the same error in the browser "[ERROR] FATAL UNHANDLED EXCEPTION: System.DllNotFoundException: libSkiaSharp".

EDIT3:  I got the compile to recognize it by replacing the multithreaded directory with the new libSkiaSharp but now it throws a compile error:  "[00:03:02.4902299] wasm-ld: error: --shared-memory is disallowed by libSkiaSharp.sk_manageddrawable.o because it was not compiled with 'atomics' or 'bulk-memory' features."  I am going to see if I can change the build script to allow multithreaded.

We are using Uno.Wasm.Bootstrap btw, not blazor.  And the regular SkiaSharp wasm release does work.

**Reply by BradChase2011** [reply]
ok I have it running.  The docker build-local.ps1 works great with some additions.  I had to create the directories in SkiaSharp\output\native\wasm\libSkiaSharp.a and SkiaSharp\output\native\wasm\libHarfBuzzSharp.a for each version and noticed after that it would copy the built native libs each time for me which was very nice!  Then I changed the build-local.ps1 to build them all for each version and feature.  They have to be in the correct order, I accidently did mt, simd which wouldnt copy because of the directory, but after changing them to simd,mt and such it would follow the correct directories.

I really really appreciate your help and thank you very much.

EDIT:  Side note, whats the best way to build all the libraries in one go?  Should I create multiple dockers for each version?
